velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -162,136 +242,246 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+        outputType_->childAt(projection.outputChannel),
+        outputBatchSize_,
+        operatorCtx_->pool());
+  }
+
+  numOutputRows_ = 0;
+  output_ = std::make_shared<RowVector>(
+      pool(), outputType_, nullptr, outputBatchSize_, std::move(localColumns));
+}
+
+void NestedLoopJoinProbe::evaluateJoinFilter(const RowVectorPtr& buildVector) {
+  // First step to process is to get a batch so we can evaluate the join
+  // filter.
+  auto filterInput = getNextCrossProductBatch(
+      buildVector,
+      filterInputType_,
+      filterProbeProjections_,
+      filterBuildProjections_);
+
+  if (filterInputRows_.size() != filterInput->size()) {
+    filterInputRows_.resizeFill(filterInput->size(), true);
+  }
+  VELOX_CHECK(filterInputRows_.isAllSelected());
+
+  std::vector<VectorPtr> filterResult;
+  EvalCtx evalCtx(
+      operatorCtx_->execCtx(), joinCondition_.get(), filterInput.get());
+  joinCondition_->eval(0, 1, true, filterInputRows_, evalCtx, filterResult);
+  filterOutput_ = filterResult[0];
+  decodedFilterResult_.decode(*filterOutput_, filterInputRows_);
+}
+
+RowVectorPtr NestedLoopJoinProbe::getNextCrossProductBatch(
+    const RowVectorPtr& buildVector,
+    const RowTypePtr& outputType,
+    const std::vector<IdentityProjection>& probeProjections,
+    const std::vector<IdentityProjection>& buildProjections) {
+  std::vector<VectorPtr> projectedChildren(outputType->size());
+  const auto numOutputRows = buildVector->size();
+
+  // Project columns from the build side.
   projectChildren(
-      projectedChildren, data, projections, numUnmatched, unmatchedMapping);
-  for (auto [_, outputChannel] : nullProjections) {
-    VELOX_CHECK_GT(projectedChildren.size(), outputChannel);
-    projectedChildren[outputChannel] = BaseVector::createNullConstant(
-        outputType_->childAt(outputChannel), numUnmatched, pool());
+      projectedChildren, buildVector, buildProjections, numOutputRows, nullptr);
+
+  // Wrap projections from the probe side as constants.
+  for (auto [inputChannel, outputChannel] : probeProjections) {
+    projectedChildren[outputChannel] = BaseVector::wrapInConstant(
+        numOutputRows, probeRow_, input_->childAt(inputChannel));
   }
+
   return std::make_shared<RowVector>(
-      pool(), outputType_, nullptr, numUnmatched, std::move(projectedChildren));
+      pool(), outputType, nullptr, numOutputRows, std::move(projectedChildren));
+}
+
+void NestedLoopJoinProbe::addOutputRow(
+    const RowVectorPtr& buildVector,
+    vector_size_t buildRow) {
+  // Probe side is always a dictionary; just populate the indice.
+  rawProbeIndices_[numOutputRows_] = probeRow_;
+
+  // Copy the build projections.
+  for (const auto& projection : buildProjections_) {
+    const auto& buildChild = buildVector->childAt(projection.inputChannel);
+    const auto& outputChild = output_->childAt(projection.outputChannel);
+    outputChild->copy(buildChild.get(), numOutputRows_, buildRow, 1);
```

Comment:
This will kill the performance.  Can we accumulate `buildRow`s as `CopyRange`s or `toSourceRow` and only call `copy` once per output batch?